debugger_protocol.rst

```diff
@@ -363,17 +364,29 @@ or Pythons Cmd module: ::
 
     command -a value -b value ...
 
-If a value for an option includes a space, the value needs to be
-encoded. You can encode values by encasing them in double quotes::
+6.0 Escaping Rules
+------------------
+
+If a value for an option includes a space or NULL character (``\0``), the
+value MUST be encapsulated in double quotes (``"``)::
 
     property_get -i 5 -n "$x['a b']" -d 0 -c 0 -p 0
 
-It is also possible to use escape characters in quoted strings::
+In other cases, the value MAY be encapsulated in double quotes::
+
+	property_get -i 6 -n "$x['ab']" -d 0 -c 0 -p 0
+
+Inside a double-quoted argument, the double-quote (``"``), back-slash (``\``)
+and NULL character (``\0``) MUST be escaped with a back-slash::
 
-    property_get -i 6 -n "$x[\"a b\"]" -d 0 -c 0 -p 0
+    property_get -i 7 -n "$x[\"a b\"]" -d 0 -c 0 -p 0
+
+And to mean ``$x["a`` *chr(0)* ``b\"]``::
+
+    property_get -i 8 -n "$x[\"a\0b\\\"]" -d 0 -c 0 -p 0
```

Comment:
Shouldn't it has \0 escaped? 
property_get -i 8 -n "$x[\"\\0b\\\"]" -d 0 -c 0 -p 0 "

Reply:
No, the char(0) needs to be escaped into ``\0``. If the property name contained ``\0``, it needs to be escaped to ``\\0``.

Reply:
Is it not yet implemented in Xdebug? With the fresh version from github I get "parse error in command".

PHP:
```
<?php
$a["bar"]["\x0"]["foo"] = "!!";
echo("");
```

```
<- property_get -i 23 -n "$a[\"bar\"][\"\0\"]" -d 0 -c 0 -p 0
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="property_get" transaction_id="23"><error code="1"><message><![CDATA[parse error in command]]></message></error></response>
```

Reply:
In this case, Xdebug would have sent you:

```
<- property_get -i 3 -n $a["bar"]
```

```
<response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="property_get" transaction_id="3">
    <property name="$a[&quot;bar&quot;]" fullname="$a[&quot;bar&quot;]" type="array" children="1" numchildren="1" page="0" pagesize="32">
        <property name="&#0;" fullname="$a[&quot;bar&quot;][&quot;\0&quot;]" type="array" children="1" numchildren="1">
        </property>
    </property>
</response>
```

The fullname is ``$a[&quot;bar&quot;][&quot;\0&quot;]``, which after XML decoding gives ``$a["bar"]["\0"]``.

As there is no chr(0) in there, but just the string ``\0``, you need to escape the ``\`` in double quotes:

    property_get -i 3 -n "$a[\"bar\"][\"\\0\"]"

But you can also just do:

    property_get -i 3 -n $a["bar"]["\0"]

Reply:
I've pushed a new change to document this better again. Let me know if you have any further questions, I'm also happy to do a call if that makes it easier.

Reply:
Is there any example when Xdebug will send chr(0) in property_get response?

Reply:
No, it currently does not... but that is a bug :-)

I've added two tickets for that for me to fix (hopefully, tonight):

- https://bugs.xdebug.org/view.php?id=1514
- https://bugs.xdebug.org/view.php?id=1515